### PR TITLE
feat: Add Screenshots automatically.

### PR DESCRIPTION
**Pull request by lucas-zimerman**
This WIP feature allows adding screenshots to events whenever possible.

### new Sentry options:

* AttachScreenshots: Attaches screenshots from the app to events automatically whenever possible.
* ScreenshotQuality: (Low, Medium, High) allows the user to choose the quality of the screenshots (default low for lower file sizes)
* file format: Jpeg whenever available.

### Goals

<s>* Use of native code so that porting this feature becomes easier to native SDKs.</s>
* Support iOS, Android, and <s>UWP</s> (other platforms doesn't support attaching a screenshot exclusively from the application)
* Have an event processor that can inject attachments to an event being processed.

Sample Event: https://sentry.io/organizations/sentry-sdks/discover/sentry-xamarin:970c172209f54cf3827b0f3813185764

I suggest the usage of jpeg by default due to the lower performance impact compared to png encoding (see: https://blender.stackexchange.com/questions/148231/what-image-format-encodes-the-fastest-or-at-least-faster-png-is-too-slow)

## Challenges

* Event Processors doesn't allow the insertion of attachments directly into events, requiring you to edit the scope (and with that, the screenshot may be attached to all future events and not the current one)

### CODE REVIEW

File: Src/Sentry.Xamarin/SentryXamarinOptions.cs [hunk 1]
```diff
@@ -8,6 +8,17 @@ namespace Sentry
     /// </summary>
     public class SentryXamarinOptions : SentryOptions
     {
+        /// <summary>
+        /// Attaches screenshots from the app to events automatically whenever possible.
+        /// </summary>
+        public bool AttachScreenshots { get; set; }
```
Comment (ueman) [inline Src/Sentry.Xamarin/SentryXamarinOptions.cs:14]: This should probably include a note about PII. 
There could be potentially private information in the screenshot. For example in bank or chat apps.

Comment (lucas-zimerman) [reply Src/Sentry.Xamarin/SentryXamarinOptions.cs:17]: I agree

File: Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachmentContent.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public void SetNewData(Stream stream)
+        {
+            _data = new byte[stream.Length];
+            stream.Read(_data, 0, _data.Length);
+        }
+
+        /// <summary>
+        /// Inform is the contenet was previously read.
```
Comment (bruno-garcia) [inline Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachmentContent.cs:27]:
```suggestion
        /// Inform is the content was previously read.
```

File: Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachmentContent.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        /// By calling this function you will reset the read state to false and return the current state.
+        /// </summary>
+        /// <returns>true if the content was read.</returns>
+        public bool GetWasRead()
```
Comment (bruno-garcia) [inline Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachmentContent.cs:31]: Odd to have such side effect on a method called "get"
```suggestion
        public bool ResetWasRead()
```

File: Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachmentContent.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        /// Get a stream from the attachment data.
+        /// </summary>
+        /// <returns>A Memory stream containing the data.</returns>
+        public Stream GetStream()
+        {
+            _wasRead = true;
+            return new MemoryStream(_data);
+        }
+    }
+}
```
Comment (bruno-garcia) [inline Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachmentContent.cs:48]: image
```suggestion
}

```

File: Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        public ScreenshotEventProcessor(SentryXamarinOptions options) => _options = options;
+
+        public SentryEvent? Process(SentryEvent @event)
+        {
+            try
+            {
+                if (_options.SessionLogger?.IsBackground() == false)
+                {
+                    var stream = Capture();
+                    if (stream != null)
```
Comment (bruno-garcia) [inline Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs:22]: Can't use it here?

```suggestion
                    if (stream is not null)
```

File: Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                    {
+                        /* Create a new attachment if no screenshot attachment was fond.
```
Comment (bruno-garcia) [inline Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs:24]:
```suggestion
                        /* Create a new attachment if no screenshot attachment was found.
```

File: Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                         * If there's an attachment content but it wasnt read during the last event processing
+                         * Assume it was cleared and create a new one.
+                         */
+                        if(_screenshot?.GetWasRead() != true)
```
Comment (bruno-garcia) [inline Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs:28]:
```suggestion
                        if (_screenshot?.GetWasRead() != true)
```

File: Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                }
+            }
+            catch (Exception ex)
+            {
+                _options.DiagnosticLogger?.LogError("Failed to capture a screenshot", ex);
+            }
+            return @event;
+        }
+
+        internal Stream Capture()
```
Comment (bruno-garcia) [inline Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs:47]: Internal for testing?
Otherwise best to inline this

**Review by bruno-garcia — APPROVED** [review]
We should release a preview with this

**Comment by bruno-garcia** [comment]
@lucas-zimerman no changelog entry?

No Danger? :)

**Pull request merged**